### PR TITLE
Adds 'nofollow' setting to Button block

### PR DESCRIPTION
**Pull request by bangank36**
## What?

Add the ability to toggle the `nofollow` for Button's link. 

Partial implementation of https://github.com/WordPress/gutenberg/issues/54091
Prior art for Rich text link https://github.com/WordPress/gutenberg/pull/53945

## Why?

Check https://github.com/WordPress/gutenberg/pull/53945#issue-1866939265

## How?

- Place the `nofollow` option in the `Advanced` settings area
- Update the [linkValue](https://github.com/WordPress/gutenberg/blob/0792221dd597290fe512b109a1469f2057f0dda4/packages/block-library/src/button/edit.js#L154) for persisting the Advanced settings state

## Testing Instructions

- Open Post Editor
- Add new Button block
- Add a link to the Button
- Click Edit link
- Toggle the `opensInNewTab` and `nofollow` 
- Check and confirm whether the `rel` and `target` attributes has been inserted correctly

| opensInNewWindow | nofollow | Generated                      |
| --------------------------- | --------------------- | ---------------------------------------- |
| ✅                         | ❌                   | `target="_blank" rel="noreferrer noopener"` |
| ❌                         | ✅                   | `target="" rel="nofollow"`               |
| ✅                         | ✅                   | `target="_blank" rel="noreferrer noopener nofollow"` |
| ❌                         | ❌                   | `target="" rel=""`                      |

### Testings
- Verify and update these e2e tests
   `npm run test:e2e:playwright -- test/e2e/specs/editor/blocks/links.spec.js`
   `npm run test:e2e:playwright -- test/e2e/specs/editor/blocks/buttons.spec.js`

- Insert new unit testing for updateLinkAttributes
  `npm run test:unit -- packages/block-library/src/button`

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/10071857/de3bcbc1-6bb7-45e8-b6eb-d7da7e6d974e

cc: @getdave

### CODE REVIEW

File: packages/block-editor/src/components/link-control/index.js
Comment (bangank36) [inline packages/block-editor/src/components/link-control/index.js]: I introduced this `hasNoFollowSetting` so that LinkControl can render the `nofollow` checkbox based on the assigned prop, this can help reduce the default overridden on various places, eg [inline.js](https://github.com/WordPress/gutenberg/blob/ab5a41c7979b81f0827289f1adde7f505bcc2702/packages/format-library/src/link/inline.js#L34)
Please let me know what you think @getdave?

File: packages/block-editor/src/components/link-control/index.js
Comment (alexstine) [inline packages/block-editor/src/components/link-control/index.js]: This causes a screen reader accessibility issue. No code tags in labels. Not actually sure why this is as it seems to be allowed per the HTML specification but with testing, it is an issue.

Maybe this is a better label?

```suggestion
					__( 'Add no-follow for search engines' ),
```

**Review by alexstine — COMMENTED** [review]
One note below.

File: packages/block-editor/src/components/link-control/index.js
Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: > This causes a screen reader accessibility issue. No code tags in labels

Good to know! I am okay with the alternative label, so let's ask for other's thoughts eg: @scruffian, then I will update the e2e testing for the new labels

Comment (scruffian) [reply packages/block-editor/src/components/link-control/index.js]: Makes sense to me.

Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: Addressed on the latest commit, I also updated the e2e testing for selecting the new label
image

Also I had another [question here](https://github.com/WordPress/gutenberg/pull/54110#discussion_r1312995012), do you mind to take a look as well @alexstine? Thanks

Comment (alexstine) [reply packages/block-editor/src/components/link-control/index.js]: @bangank36 Thanks. I honestly don't know about your other question. I mainly focus on UI accessibility in Gutenberg, the internals of managing settings is an unknown area to me. It sounds like it makes sense but that's really all I can offer.

Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: This will need updating in `trunk` for Rich Text as well.

Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: You mean for the [inline.js](https://github.com/WordPress/gutenberg/blob/ab5a41c7979b81f0827289f1adde7f505bcc2702/packages/format-library/src/link/inline.js#L34) file? 
https://github.com/WordPress/gutenberg/blob/8ae150433d58ffbb0536f4bc1435b043b905b0a0/packages/format-library/src/link/inline.js#L38

Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: Yes. I can't recall whether I used a `code` tag there or not in the end.

If I did then we'll need a separate PR to fix that one.

Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: See https://github.com/WordPress/gutenberg/issues/54179

Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: In https://github.com/WordPress/gutenberg/pull/53945 you used `<code>` tag, and I changed the inline.js and the e2e testings for link with the suggested label already ( along with the button link changes )
If you think they need to be on its own PR then I will submit a new commit to remove these changes from this PR.
Also I may need your thoughts on https://github.com/WordPress/gutenberg/pull/54110#discussion_r1312995012

Thanks

Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: > If you think they need to be on its own PR then I will submit a new commit to remove these changes from this PR.

Whilst it seems over the top, I would make a separate PR. Why? Because if _this_ PR gets rejected or takes a while to land the a11y bug will continue to exist in `trunk`. With a separate PR we can quickly and easily revert the issue immediately. It's also a better separate of concerns.

Given that you have the code already, would you be open to raising that PR?

Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: On it!

Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: See https://github.com/WordPress/gutenberg/pull/54184, I will rebase this to `trunk` once it is merged

**Comment by scruffian** [comment]
This is looking good, just needs a rebase I think.

**Review by getdave — CHANGES_REQUESTED** [review]
I'd like some time to check [this](https://github.com/WordPress/gutenberg/pull/54110#discussion_r1312995012) out thoroughly before this is merged.

File: packages/block-editor/src/components/link-control/index.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: I'm not sure we need/want to do this.

The API of this component is already extremely large and continuing to add to it only compounds the confusion.

We already have a means of determining whether or not to render settings fields. I would prefer instead if we perhaps considering exposing some other standardised settings as part of `LinkControl` in a simple fashion as we do for [the default settings](https://github.com/WordPress/gutenberg/blob/ab5a41c7979b81f0827289f1adde7f505bcc2702/packages/format-library/src/link/inline.js#L35).

That way consumers could optionally utilise those standard settings without using having to further complicate the API of this component.

My thinking is influecned by the thought that whilst `nofollow` might relevant _today_ it could easily not be in future. If we make it a standard of the component's API then we're committing to maintaining it forever. What happens if Google says "we no longer recommend this" and then we've added this prop? I'd rather avoid that.

Whilst I'm conscious that my proposal is perhaps less flexible and requires more wiring, it helps to limit the surface area of backwards compatibility issues.

What do you think?

**Review by getdave — DISMISSED** [review]
I'm going to block this PR because I"m not confident that we should be adding to the API of this component.

Let's discuss this further and get a wider range of opinions before we proceed. 

Thank you 🙇

File: packages/format-library/src/link/inline.js
Comment (getdave) [inline packages/format-library/src/link/inline.js]:
```suggestion
		title: __( 'Mark as nofollow for search engines' ),
```

This shouldn't be changing in this PR.

Comment (getdave) [reply packages/format-library/src/link/inline.js]: Let's rebase and avoid making changes to `inline.js` here 🙇

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Whist potentially "cleaner" I found this code much harder to read and understand than what we have in `inline.js`.

I wonder whether the simplest approach would be easier to follow in future?

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Why `undefined` here?

File: packages/block-editor/src/components/link-control/index.js
Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: The points make sense to me 👍 
So for now we just follow the way `inline.js` is using to append new settings to the link

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Unlike the inline link editor, the `rel` attr of Button link is editable, therefore when updating `rel` using LinkControl, we have to take the current `rel` attribute value into account (like skipping insert when existing, removing when the checkbox is false)

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: As explain [here](https://github.com/WordPress/gutenberg/pull/54110#discussion_r1319295977), since I used string concatenation, this line [let updatedRel = rel || '';](https://github.com/WordPress/gutenberg/blob/8ae150433d58ffbb0536f4bc1435b043b905b0a0/packages/block-library/src/button/edit.js#L88)  to avoid result like `undefined nofollow`

And in case `updatedRel` value is empty in the end, converting to `undefined` will ensure the HTML attribute `rel=''` is not inserted in the block HTML, am I too inflexible on this, or should we just update e2e testing snapshot to match the case?

File: packages/format-library/src/link/inline.js
Comment (bangank36) [reply packages/format-library/src/link/inline.js]: Sorted!

File: packages/block-library/src/button/edit.js
Comment (getdave) [reply packages/block-library/src/button/edit.js]: Ah I see yes. Thanks for that explanation 👍

File: packages/block-editor/src/components/link-control/index.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/index.js]:
```suggestion
```

File: packages/block-editor/src/components/link-control/index.js
Comment (getdave) [inline packages/block-editor/src/components/link-control/index.js]:
```suggestion
```

File: packages/block-editor/src/components/link-control/index.js
Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js]: That sounds good to me for now. We can always refactor later.

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]:
```suggestion
```

File: packages/block-editor/src/components/link-control/index.js
Comment (bangank36) [reply packages/block-editor/src/components/link-control/index.js]: Addressed in latest commit https://github.com/WordPress/gutenberg/pull/54110/commits/07cee9c9dea8b765070bd2ca4665c7a91c0c4fa0

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Could we combine the concept of updating the entire link into the `updateLinkAttributes` method? That will bring it into line with other implementations. This will help pave the way for a future refactoring I'm planning.

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (getdave) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: You could do something like this instead to make the test more readable:

https://github.com/WordPress/gutenberg/blob/b05bee7e2e8eefb2be06c228002e4f9783847747/test/e2e/specs/editor/blocks/links.spec.js#L74-L82

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (getdave) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: You've already got selectors for these above. Let's reuse them rather than re-query.

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (getdave) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: There's no tabbing going on here...

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (getdave) [inline test/e2e/specs/editor/blocks/buttons.spec.js]: Same comment as above in https://github.com/WordPress/gutenberg/pull/54110/files#r1324188149

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Sure! Unfortunately I still can't figure an optimized way for this https://github.com/WordPress/gutenberg/pull/54110#discussion_r1318554127 yet 😅

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (bangank36) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Thanks for the suggestion, I was able to find some more useful tips for writing e2e tests at [overusing-snapshots.md](https://github.com/WordPress/gutenberg/blob/21b6a3717c1b09c376f7a5d914b0953aa2cdcddd/docs/contributors/code/e2e/overusing-snapshots.md#L4)

I also notice the `<!-- wp:buttons -->` test across the file and other block tests like `list`, If you have plan to refactor these as well, feel free to count me in when the plan is elaborated

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (bangank36) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Good spot!

File: test/e2e/specs/editor/blocks/buttons.spec.js
Comment (bangank36) [reply test/e2e/specs/editor/blocks/buttons.spec.js]: Sorted

File: packages/block-library/src/button/edit.js
Comment (getdave) [reply packages/block-library/src/button/edit.js]: I think it's fine then. Perhaps add a comment to explain that `rel` is editable as plaintext.

File: packages/block-library/src/button/edit.js
Comment (getdave) [reply packages/block-library/src/button/edit.js]: I think the optimisation is ok as it is now you've explained it.

Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Ah okay, thanks

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Something like https://github.com/WordPress/gutenberg/pull/54110/files#diff-e335a5733e9910d51f3adb7c21b371b738fdc190e6b9b71b19ba2ad45703fea1R107

**Comment by bangank36** [comment]
Hi @getdave, just a friendly nudge to request your review of the latest changes on my PR when you have a moment. Thanks!

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js:42]: Note to self that we should expose these settings as non-defaults from the component.

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: I feel this function is sufficiently complex to warrant a small set of unit tests. These can live alongside the block in the `test` folder in a file called `update-link-attributes.js`.

The tests should validate that the new value is correctly computed given various values that could be pased to it.

I'm keen that we test the regex.

**Review by getdave — COMMENTED** [review]
This is very close. Thank you for your patience.

I think upon reflection some unit tests would help add confidence here. They will also lay a solid foundation for a later refactoring.

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Good idea! I'll give it a try

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js:42]: I moved those values to ~~/utils~~ `/constants` file

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: @getdave, since `updateLinkAttributes` is defined inside the functional component and can’t be accessed from outside the component. Therefore my approach would be moving the logic into a separate function in `/utils.js`, so we can test that the attributes generation is correct. Please let me know your thoughts. Thanks 🙇 

You can run the test by `npm run test:unit -- packages/block-library/src/button`

File: packages/block-library/src/button/utils.js
Comment (getdave) [inline packages/block-library/src/button/utils.js]: Nit: Let's move this to a dedicated file `update-link-attributes.js`.

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Let's make a `constants.js` file and move these here. We can then import as needed.

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Let's move to a dedicated file name as per the function

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js]: Reading this I think we should call the function something else.

Perhaps `getUpdatedLink`?

**Review by getdave — COMMENTED** [review]
Great work here. A few more tweaks and we should be in a good place.

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: I think you are right, the current name `updateLinkAttributes` would make sense if we change its definitions like 

```diff
- export const updateLinkAttributes = ( { rel = '', url = '', opensInNewTab, nofollow } ) => {
+ export const updateLinkAttributes = ( { rel = '', url = '', opensInNewTab, nofollow, setAttributes } ) => {
  ...
-  return {
-        url: prependHTTP( url ),
-        linkTarget: newLinkTarget,
-        rel: updatedRel || undefined,
-   };
+  return () => {
+    setAttributes( {
+        url: prependHTTP( url ),
+        linkTarget: newLinkTarget,
+        rel: updatedRel || undefined,
+    } );
+  };
};
```

therefore the name changed to `getUpdatedLinkAttributes` as suggested

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Sorted

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js]: Sorted

File: packages/block-library/src/button/utils.js
Comment (bangank36) [reply packages/block-library/src/button/utils.js]: Moved to its dedicated file

File: packages/block-library/src/button/edit.js
Comment (getdave) [reply packages/block-library/src/button/edit.js]: I dislike passing `setAttributes` around.

I prefer to see this function as _preparing_ the new link. Where and how the attributes are persisted is a separate issue and we should separate those concerns. So let's not pass `setAttributes`.

Your work now is super valuable for the future of the link management in Gutenberg. I am working on a PR now that will allow all this to make sense. 

Thank you 🙇

Comment (bangank36) [reply packages/block-library/src/button/edit.js]: > I dislike passing setAttributes around.

@getdave Agree, it's just like a self-note for explaining the function name changed, it is `getUpdatedLinkAttributes` already from the latest commit

File: packages/block-library/src/button/edit.js
Comment (getdave) [inline packages/block-library/src/button/edit.js:256]: Question: do we need these aliases? I'd like to contain all the logic for computing the new link value in the function rather than leaving anything outside if possible.

**Review by getdave — APPROVED** [review]
Thank you for all the work on this.

LGTM 🚀

File: packages/block-library/src/button/edit.js
Comment (bangank36) [reply packages/block-library/src/button/edit.js:256]: I used aliases because lint-js is not happy with duplicate `url` variable name, am I missing something here? 

image

https://github.com/WordPress/gutenberg/blob/f6e49a7a8e4db6c3bdbbc033779bf88db1c90175/packages/block-library/src/button/edit.js#L100

Comment (getdave) [reply packages/block-library/src/button/edit.js:256]: Ah I see. I've seen this alias pattern used to ensure default values before, but I just realised the alias would need to have default values for that to be the case.

All good as is 👍

Comment (bangank36) [reply packages/block-library/src/button/edit.js:256]: Great! Thanks for all the dedicated reviews and suggestions

**Comment by getdave** [comment]
Note to WP 6.4 leads. I don't believe we need to backport this PR. Happy to be advised otherwise if folks find it to be sufficiently "blessed".

**Pull request merged**
